src/wasm/wasm-type.cpp

```diff
@@ -103,38 +172,79 @@ void Type::init(const std::vector<Type>& types) {
     *this = types[0];
     return;
   }
+  id = canonicalize(TypeDef(tuple));
+}
 
-  // Add a new type if it hasn't been added concurrently
-  std::lock_guard<std::mutex> lock(mutex);
-  auto indexIt = indices.find(types);
-  if (indexIt != indices.end()) {
-    id = indexIt->second;
+Type::Type(const Signature signature, bool nullable) {
+  id = canonicalize(TypeDef(signature, nullable));
+}
+
+Type::Type(const Struct& struct_, bool nullable) {
+#ifndef NDEBUG
+  for (Field f : struct_.fields) {
+    assert(f.type.isSingle() && f.type.isConcrete());
+  }
+#endif
+  id = canonicalize(TypeDef(struct_, nullable));
+}
+
+Type::Type(const Array& array, bool nullable) {
+#ifndef NDEBUG
+  assert(array.element.type.isSingle() && array.element.type.isConcrete());
+#endif
+  id = canonicalize(TypeDef(array, nullable));
+}
+
+bool Type::isTuple() const {
+  if (id <= _last_basic_id) {
+    return false;
   } else {
-    auto vec = std::make_unique<std::vector<Type>>(types);
-    id = uintptr_t(vec.get());
-    constructedTypes.push_back(std::move(vec));
-    assert(id > _last_value_type);
-    indices[types] = id;
+    auto* typeDef = (TypeDef*)id;
+    return typeDef->isTuple();
   }
 }
 
-Type::Type(std::initializer_list<Type> types) { init(types); }
+bool Type::isRef() const {
+  if (id <= _last_basic_id) {
+    return id >= funcref && id <= exnref;
+  } else {
+    auto* typeDef = (TypeDef*)id;
+    switch (typeDef->kind) {
+      case TypeDef::TupleKind:
+        return false;
+      case TypeDef::SignatureKind:
+      case TypeDef::StructKind:
+      case TypeDef::ArrayKind:
+        return true;
+    }
+    WASM_UNREACHABLE("unexpected kind");
+  }
+}
 
-Type::Type(const std::vector<Type>& types) { init(types); }
+bool Type::isNullable() const {
+  if (id <= _last_basic_id) {
+    return id >= funcref && id <= exnref;
```

Comment:
Is [`exnref` nullable](https://github.com/WebAssembly/exception-handling/issues/90)?

Reply:
yep! All reference types introduced before the `let` instruction are nullable because they need to be stored in locals and those locals need to have default null values.